pkg/nfs/controllerserver.go

```diff
@@ -73,3 +161,177 @@ func (cs *ControllerServer) ListSnapshots(ctx context.Context, req *csi.ListSnap
+	sharePath := filepath.Join("/" + vol.baseDir)
+	targetPath := cs.getInternalMountPath(vol)
+	stdVolCap := csi.VolumeCapability{
+		AccessType: &csi.VolumeCapability_Mount{
+			Mount: &csi.VolumeCapability_MountVolume{},
+		},
+	}
+
+	glog.V(4).Infof("internally mounting %v:%v at %v", vol.server, sharePath, targetPath)
+	_, err := cs.Driver.ns.NodePublishVolume(ctx, &csi.NodePublishVolumeRequest{
+		TargetPath: targetPath,
+		VolumeContext: map[string]string{
+			"server": vol.server,
+			"share":  sharePath,
+		},
+		VolumeCapability: &stdVolCap,
+		VolumeId:         vol.id,
+	})
+	return err
+}
+
+// Unmount nfs server at base-dir
+func (cs *ControllerServer) internalUnmount(ctx context.Context, vol *nfsVolume) error {
+	targetPath := cs.getInternalMountPath(vol)
+
+	// Unmount nfs server at base-dir
+	glog.V(4).Infof("internally unmounting %v", targetPath)
+	_, err := cs.Driver.ns.NodeUnpublishVolume(ctx, &csi.NodeUnpublishVolumeRequest{
+		TargetPath: cs.getInternalMountPath(vol),
+	})
+	return err
+}
+
+// Convert VolumeCreate parameters to an nfsVolume
+func (cs *ControllerServer) newNFSVolume(name string, params map[string]string) (*nfsVolume, error) {
+	var (
+		server  string
+		baseDir string
+	)
+
+	// Validate parameters (case-insensitive).
+	// TODO do more strict validation.
+	for k, v := range params {
+		switch strings.ToLower(k) {
+		case paramServer:
+			server = v
+		case paramBaseDir:
+			baseDir = v
+		default:
+			return nil, fmt.Errorf("invalid parameter %q", k)
+		}
+	}
+
+	// Validate required parameters
+	if server == "" {
+		return nil, fmt.Errorf("%v is a required parameter", paramServer)
+	}
+	if baseDir == "" {
+		return nil, fmt.Errorf("%v is a required parameter", paramBaseDir)
+	}
+
+	vol := &nfsVolume{
+		server:  server,
+		baseDir: baseDir,
+		subDir:  name,
+	}
+	vol.id = cs.getVolumeIdFromNfsVol(vol)
+
+	return vol, nil
+}
+
+// Get working directory for CreateVolume and DeleteVolume
+func (cs *ControllerServer) getInternalMountPath(vol *nfsVolume) string {
+	return filepath.Join(cs.workingMountDir, vol.subDir)
+}
+
+// Get internal path where the volume is created
+// The reason why the internal path is "workingDir/subDir/subDir" is because:
+//   * the semantic is actually "workingDir/volId/subDir" and volId == subDir.
+//   * we need a mount directory per volId because you can have multiple
+//     CreateVolume calls in parallel and they may use the same underlying share.
+//     Instead of refcounting how many CreateVolume calls are using the same
+//     share, it's simpler to just do a mount per request.
+func (cs *ControllerServer) getInternalVolumePath(vol *nfsVolume) string {
+	return filepath.Join(cs.getInternalMountPath(vol), vol.subDir)
+}
+
+// Get user-visible share path for the volume
+func (cs *ControllerServer) getVolumeSharePath(vol *nfsVolume) string {
+	return filepath.Join("/", vol.baseDir, vol.subDir)
```

Comment:
why this `filepath.Join` takes 3 parameters?

Reply:
This is the path of the share that was created and put into the returned VolumeContext. ie, if you have Storageclass param `base-dir = myvol`, `server = 10.10.10.10`, then after CreateVolume finishes, the VolumeContext returned should be: `server = 10.10.10.10`, `share = /myvol/foo-subdir`